src/runtime_src/core/common/api/xclbin_int.h

```diff
@@ -45,6 +35,110 @@ is_valid_or_error(xrtXclbinHandle handle);
+ *
+ * @my_xclbin:  Xclbin object to operate on
+ * @section:    The desired section
+ * @xclbin_id:  The uuid that corresponds to my_xclbin
+ * Return:      Pair of section data and size in bytes
+ */
+std::pair<const char*, size_t>
+get_axlf_section(const ::xrt::xclbin& my_xclbin, axlf_section_kind section, const uuid& xclbin_id = uuid());
+
+/**
+ * get_axlf_section() - Get section from currently loaded axlf, throws if section not found
+ *
+ * @my_xclbin:  Xclbin object to operate on
+ * @section:    The desired section
+ * @xclbin_id:  The uuid that corresponds to my_xclbin
+ * Return:      Pair of section data and size in bytes
+ */
+std::pair<const char*, size_t>
+get_axlf_section_or_error(const ::xrt::xclbin& my_xclbin, axlf_section_kind section, const uuid& xclbin_id = uuid());
+
+/**
+ * get_section_kinds() - Iterates over enum
+ *
+ * Return:     Vector of all possible section kinds
+ *
+ */
+std::vector<axlf_section_kind>
+get_section_kinds();
+
+/**
+ * create_xclbin() - Creates an xclbin object that only caches specified sections
+ *  by default all present sections are cached
+ *
+ * @top:       raw data of xclbin
+ * @kinds:     list of axlf_section_kind to cache
+ * Return:     xclbin object
+ *
+ */
+::xrt::xclbin
+create_xclbin(const axlf* top, const std::vector<axlf_section_kind>& kinds = get_section_kinds());
+
+/**
+ * create_xclbin() - creates a xclbin object
+ *  by default all present sections are cached
+ *
+ * @device:    ptr to xrt_core::device
+ * @kinds:     list of axlf_section_kind to cache
+ * Return:     xclbin object
+ *
+ */
+::xrt::xclbin
+create_xclbin(const xrt_core::device* device, const std::vector<axlf_section_kind>& kinds = get_section_kinds());
+
+/**
+ * read_xclbin() - Reads xclbin file as binary
+ *
+ * Return:     Vector of char
+ *
+ */
+std::vector<char>
+read_xclbin(const std::string& fnm);
+
+/**
+ * Utility to get uuid from xclbin -
+ */
+uuid
+get_uuid(const axlf* top);
+
+/**
+ * header_section_iterator: iterator to traverse axlf_section_header
+ */
+class header_section_iterator // const behavior intended
+{
+public:
+  explicit header_section_iterator(const axlf_section_header* ptr) : m_ptr(const_cast<axlf_section_header*>(ptr)) {}
+  header_section_iterator operator++() { auto i = *this; m_ptr++; return i; }
+  header_section_iterator operator++(int) { m_ptr++; return *this; }
+  const axlf_section_header& operator*() { return *m_ptr; }
+  const axlf_section_header* operator->() { return m_ptr; }
+  bool operator==(const header_section_iterator& rhs) { return m_ptr == rhs.m_ptr; }
+  bool operator!=(const header_section_iterator& rhs) { return m_ptr != rhs.m_ptr; }
+private:
+  axlf_section_header* m_ptr;
+};
+
+/**
+ * axlf_header_container: header section container for axlf
+ */
+class axlf_header_container
+{
+public:
+  explicit axlf_header_container(const axlf* top) : m_begin(top->m_sections), m_end(&(top->m_sections[top->m_header.m_numSections])) {}
```

Comment:
Let's do `m_end(top->m_sections + top->m_header.m_numSections)`  This is more std'ish.